File: backend/services/four_column_daybook.py
from sqlalchemy.orm import Session
from datetime import date

from models.four_column_daybook import FourColumnDayBook
from schemas.four_column_daybook import FourColumnDayBookCreate

from models.account import Account

from models.traders_due_report import TradersDueReport
from models.farmers_due_report import FarmersDueReport



# =====================================================
# CREATE
# =====================================================

def create_daybook_entry(
    db: Session,
    entry: FourColumnDayBookCreate
):

    # ==========================================
    # SAVE IN FOUR COLUMN DAY BOOK
    # ==========================================

    new_entry = FourColumnDayBook(

        transaction_no=entry.transaction_no,

        voucher_no=entry.voucher_no,

        date=entry.date,

        voucher_type=entry.voucher_type,

        group_name=entry.group_name,

        income=entry.income,

        credit=entry.credit,

        account_name=entry.account_name,

        pb_no=entry.pb_no,

        expenditure=entry.expenditure,

        narration=entry.narration,

        bill_no=entry.bill_no,

        amount=entry.amount,

        running_total=entry.running_total

    )


    db.add(new_entry)

    db.commit()

    db.refresh(new_entry)



    # ==========================================
    # IF TRADERS, SAVE IN TRADERS DUE REPORT
    # ==========================================

    if entry.group_name == "Traders":

        trader_account = db.query(
            Account
        ).filter(

            Account.account_name == entry.account_name,

            Account.group_name == "Traders"

        ).first()


        town = ""

        if trader_account:

            town = trader_account.town



        traders_due = TradersDueReport(

            town=town,

            account_name=entry.account_name,

            bill_date=entry.date,

            bill_no=entry.bill_no,

            bill_amount=entry.amount

        )


        db.add(traders_due)

        db.commit()



    # ==========================================
    # IF FARMERS, SAVE IN FARMERS DUE REPORT
    # ==========================================

    if entry.group_name == "Farmers":

        farmer_account = db.query(
            Account
        ).filter(

            Account.account_name == entry.account_name,

            Account.group_name == "Farmers"

        ).first()


        town = ""

        if farmer_account:

            town = farmer_account.town



        farmers_due = FarmersDueReport(

            town=town,

            account_name=entry.account_name,

            bill_date=entry.date,

            bill_no=entry.bill_no,

            bill_amount=entry.amount

        )


        db.add(farmers_due)

        db.commit()



    return new_entry





# =====================================================
# GET ALL
# =====================================================

def get_all_daybook_entries(
    db: Session
):

    return db.query(

        FourColumnDayBook

    ).order_by(

        FourColumnDayBook.id.desc()

    ).all()





# =====================================================
# GET BY ID
# =====================================================

def get_daybook_entry(

    db: Session,

    entry_id: int

):

    return db.query(

        FourColumnDayBook

    ).filter(

        FourColumnDayBook.id == entry_id

    ).first()





# =====================================================
# UPDATE
# =====================================================

def update_daybook_entry(

    db: Session,

    entry_id: int,

    entry: FourColumnDayBookCreate

):

    existing_entry = db.query(

        FourColumnDayBook

    ).filter(

        FourColumnDayBook.id == entry_id

    ).first()



    if not existing_entry:

        return None



    existing_entry.transaction_no = entry.transaction_no

    existing_entry.voucher_no = entry.voucher_no

    existing_entry.date = entry.date

    existing_entry.voucher_type = entry.voucher_type

    existing_entry.group_name = entry.group_name

    existing_entry.income = entry.income

    existing_entry.credit = entry.credit

    existing_entry.account_name = entry.account_name

    existing_entry.pb_no = entry.pb_no

    existing_entry.expenditure = entry.expenditure

    existing_entry.narration = entry.narration

    existing_entry.bill_no = entry.bill_no

    existing_entry.amount = entry.amount

    existing_entry.running_total = entry.running_total



    db.commit()

    db.refresh(existing_entry)

    return existing_entry





# =====================================================
# VOUCHER TYPE BASE NUMBERS
# =====================================================

VOUCHER_START_NUMBERS = {

    "Cash Payment": 1,

    "Cash Receipt": 100,

    "Bank Payment": 200,

    "Bank Receipt": 300,

    "Contra": 400,

    "Journal": 500

}





# =====================================================
# NEXT BILL NUMBER
# =====================================================

def get_next_bill_no(

    db: Session

):

    last_bill = db.query(

        FourColumnDayBook

    ).order_by(

        FourColumnDayBook.bill_no.desc()

    ).first()



    if not last_bill:

        return 1000



    return int(last_bill.bill_no) + 1





# =====================================================
# NEXT TRANSACTION NUMBER
# =====================================================

def get_next_transaction_no(

    db: Session

):

    last_transaction = db.query(

        FourColumnDayBook

    ).order_by(

        FourColumnDayBook.transaction_no.desc()

    ).first()



    if not last_transaction:

        return 1000



    return int(last_transaction.transaction_no) + 1





# =====================================================
# NEXT VOUCHER NUMBER
# =====================================================

def get_next_voucher_no(

    db: Session,

    voucher_type: str

):

    return VOUCHER_START_NUMBERS.get(

        voucher_type,

        1

    )





# =====================================================
# NEW ENTRY DEFAULTS
# =====================================================

def get_new_daybook_defaults(

    db: Session,

    voucher_type: str

):

    return {

        "transaction_no": get_next_transaction_no(db),

        "voucher_no": get_next_voucher_no(db, voucher_type),

        "date": date.today(),

        "voucher_type": voucher_type,

        "bill_no": get_next_bill_no(db)

    }





# =====================================================
# DELETE
# =====================================================

def delete_daybook_entry(

    db: Session,

    entry_id: int

):

    existing_entry = db.query(

        FourColumnDayBook

    ).filter(

        FourColumnDayBook.id == entry_id

    ).first()



    if not existing_entry:

        return None



    # ==========================================
    # DELETE FARMERS DUE REPORT
    # ==========================================

    if existing_entry.group_name == "Farmers":

        db.query(

            FarmersDueReport

        ).filter(

            FarmersDueReport.account_name == existing_entry.account_name,

            FarmersDueReport.bill_no == existing_entry.bill_no

        ).delete()



    # ==========================================
    # DELETE TRADERS DUE REPORT
    # ==========================================

    if existing_entry.group_name == "Traders":

        db.query(

            TradersDueReport

        ).filter(

            TradersDueReport.account_name == existing_entry.account_name,

            TradersDueReport.bill_no == existing_entry.bill_no

        ).delete()



    # ==========================================
    # DELETE FOUR COLUMN DAY BOOK
    # ==========================================

    db.delete(existing_entry)

    db.commit()


    return existing_entry

File: my-app/src/transactions/FourColumnDayBook.js
import React, { useEffect, useState } from "react";
import "./FourColumnDayBook.css";
import API from "../api";

function FourColumnDayBook() {

  const voucherTypes = [
    "Cash Payment",
    "Cash Receipt",
    "Bank Payment",
    "Bank Receipt",
    "Contra",
    "Journal"
  ];

  const createEmptyRows = () => {

    return Array.from(
      { length: 15 },
      () => ({

        income: "",
        credit: "",
        account_name: "",
        pb_no: "",
        expenditure: "",
        narration: ""

      })
    );

  };


  const [transactionNo, setTransactionNo] =
    useState("");

  const [voucherNo, setVoucherNo] =
    useState("");

  const [date, setDate] =
    useState("");

  const [billNo, setBillNo] =
    useState("");

  const [amount, setAmount] =
    useState("");

  const [runningTotal, setRunningTotal] =
    useState("0.00");


  const [selectedVoucherType, setSelectedVoucherType] =
    useState("Cash Payment");


  const [rows, setRows] =
    useState(createEmptyRows());
    const [selectedGroup, setSelectedGroup] = useState("All Accs");

const [accounts, setAccounts] = useState([]);


  // =====================================================
  // LOAD DEFAULTS
  // =====================================================

  useEffect(() => {

    loadDefaults(selectedVoucherType);

  }, [selectedVoucherType]);

  useEffect(() => {

  loadAccounts(selectedGroup);

}, [selectedGroup]);


  const loadDefaults = async (voucherType) => {

    try {

      const response = await fetch(

        `${API}/four-column-daybook/new/defaults?voucher_type=${encodeURIComponent(
          voucherType
        )}`

      );


      if (!response.ok) {

        throw new Error(
          "Failed to load defaults"
        );

      }


      const data =
        await response.json();


      setTransactionNo(
        data.transaction_no
      );


      setVoucherNo(
        data.voucher_no
      );


      setDate(
        data.date
      );


      setBillNo(
        data.bill_no
      );


    }

    catch (error) {

      console.error(
        "Error loading defaults:",
        error
      );

    }

  };

  const loadAccounts = async (group) => {

  try {

    let url =
      `${API}/accounts`;

    if (group !== "All Accs") {

      url =
       `${API}/accounts/group/${group}`;

    }

    const response =
      await fetch(url);

    const data =
      await response.json();

    setAccounts(data);

  }

  catch (error) {

    console.error(error);

  }

};


  // =====================================================
  // CHANGE ROW DATA
  // =====================================================

  const handleRowChange = (

    index,
    field,
    value

  ) => {


    const updatedRows =
      [...rows];


    updatedRows[index] = {

      ...updatedRows[index],

      [field]: value

    };


    setRows(
      updatedRows
    );

  };


  // =====================================================
  // CALCULATE TOTAL
  // =====================================================

  const handleCalculate = () => {


    let total = 0;


    rows.forEach((row) => {


      total +=
        parseFloat(row.income) || 0;


      total +=
        parseFloat(row.credit) || 0;


      total +=
        parseFloat(row.expenditure) || 0;


    });


    total +=
      parseFloat(amount) || 0;


    setRunningTotal(
      total.toFixed(2)
    );

  };


  // =====================================================
  // SAVE ONLY ENTERED ROWS
  // =====================================================

  const handleSave = async () => {


    /*
      ONLY ROWS WITH DATA WILL BE SAVED.

      EMPTY ROWS WILL NOT BE SENT
      TO THE BACKEND.
    */


    const enteredRows =
      rows.filter((row) => {


        return (

          row.income !== "" ||

          row.credit !== "" ||

          row.account_name.trim() !== "" ||

          row.pb_no.trim() !== "" ||

          row.expenditure !== "" ||

          row.narration.trim() !== ""

        );

      });


    // ================================================
    // NO DATA CHECK
    // ================================================

    if (
      enteredRows.length === 0
    ) {

      alert(
        "Please enter at least one transaction row"
      );

      return;

    }


    // ================================================
    // ACCOUNT NAME CHECK
    // ================================================

    const invalidRow =
      enteredRows.find(

        (row) =>
          row.account_name.trim() === ""

      );


    if (invalidRow) {

      alert(
        "Please enter Account Name for every entered row"
      );

      return;

    }


    try {


      /*
        SAVE EACH ENTERED ROW

        Example:

        15 rows displayed

        User enters data in:

        Row 1
        Row 2
        Row 5

        Only 3 records are sent
        to the database.
      */


      for (
        const row
        of enteredRows
      ) {


        const response =
          await fetch(

            `${API}/four-column-daybook/`,

            {

              method: "POST",

              headers: {

                "Content-Type":
                  "application/json"

              },


              body: JSON.stringify({

                transaction_no:
                  String(transactionNo),


                voucher_no:
                  String(voucherNo),


                date:
                  date,


                voucher_type:
                  selectedVoucherType,

                    group_name: selectedGroup,  


                income:
                  Number(row.income) || 0,


                credit:
                  Number(row.credit) || 0,


                account_name:
                  row.account_name.trim(),


                pb_no:
                  row.pb_no.trim() || null,


                expenditure:
                  Number(row.expenditure) || 0,


                narration:
                  row.narration.trim() || null,


                bill_no:
                  String(billNo),


                amount:
                  Number(amount) || 0,


                running_total:
                  Number(runningTotal) || 0

              })

            }

          );


        if (
          !response.ok
        ) {


          const errorData =
            await response.json();


          console.error(
            "Backend error:",
            errorData
          );


          throw new Error(
            JSON.stringify(errorData)
          );

        }

      }


      // ================================================
      // SUCCESS
      // ================================================

      alert(
        "Day book entry saved successfully"
      );


      // ================================================
      // CLEAR ROWS
      // ================================================

      setRows(
        createEmptyRows()
      );


      setAmount(
        ""
      );


      setRunningTotal(
        "0.00"
      );


      // ================================================
      // LOAD NEXT NUMBERS
      // ================================================

      loadDefaults(
        selectedVoucherType
      );


    }

    catch (error) {


      console.error(
        "Error saving daybook entry:",
        error
      );


      alert(
        "Error saving daybook entry. Check console."
      );

    }

  };


  // =====================================================
  // CLEAR
  // =====================================================

  const handleClear = () => {


    setRows(
      createEmptyRows()
    );


    setAmount(
      ""
    );


    setRunningTotal(
      "0.00"
    );

  };


  return (

    <div className="daybook-container">


      <h2 className="daybook-title">

        Four Column Day Book Entry

      </h2>


      {/* =====================================================
          HEADER
      ===================================================== */}

      <div className="header-card">


        <div className="header-grid">


          <div className="field">

            <label>
              Transaction No
            </label>


            <input

              type="text"

              value={
                transactionNo
              }

              readOnly

            />

          </div>


          <div className="field">

            <label>
              Voucher No
            </label>


            <input

              type="text"

              value={
                voucherNo
              }

              readOnly

            />

          </div>


          <div className="field">

            <label>
              Date
            </label>


            <input

              type="date"

              value={
                date
              }

              readOnly

            />

            <div className="field">

  <label>All Accs</label>

  <select

    value={selectedGroup}

    onChange={(e) =>
      setSelectedGroup(e.target.value)
    }

  >

    <option>All Accs</option>
    <option>Traders</option>
    <option>Farmers</option>
    
    

  </select>

</div>

          </div>


        </div>


      </div>


      {/* =====================================================
          VOUCHER TABS
      ===================================================== */}

      <div className="voucher-tabs">


        {voucherTypes.map(

          (voucherType) => (


            <button

              key={
                voucherType
              }

              className={

                selectedVoucherType ===
                voucherType

                  ? "active-tab"

                  : ""

              }


              onClick={() =>

                setSelectedVoucherType(
                  voucherType
                )

              }

            >

              {
                voucherType
              }

            </button>

          )

        )}


      </div>


      {/* =====================================================
          TRANSACTION TABLE
      ===================================================== */}

      <div className="table-card">


        <div className="table-wrapper">


          <table className="daybook-table">


            <thead>


              <tr>

                <th>
                  Income
                </th>

                <th>
                  Credit
                </th>

                <th>
                  Account Name
                </th>

                <th>
                  P.B No
                </th>

                <th>
                  Expenditure
                </th>

                <th>
                  Narration
                </th>

              </tr>


            </thead>


            <tbody>


              {rows.map(

                (row, index) => (


                  <tr
                    key={index}
                  >


                    <td>

                      <input

                        type="number"

                        placeholder="0.00"

                        value={
                          row.income
                        }


                        onChange={(e) =>

                          handleRowChange(

                            index,

                            "income",

                            e.target.value

                          )

                        }

                      />

                    </td>


                    <td>

                      <input

                        type="number"

                        placeholder="0.00"

                        value={
                          row.credit
                        }


                        onChange={(e) =>

                          handleRowChange(

                            index,

                            "credit",

                            e.target.value

                          )

                        }

                      />

                    </td>


                    <td>

                     <select

  value={row.account_name}

  onChange={(e) =>

    handleRowChange(

      index,

      "account_name",

      e.target.value

    )

  }

>

  <option value="">
    Select Account
  </option>

  {accounts.map((account) => (

    <option

      key={account.id}

      value={account.account_name}

    >

      {account.account_name}

    </option>

  ))}

</select>

                    </td>


                    <td>

                      <input

                        type="text"

                        placeholder="P.B No"

                        value={
                          row.pb_no
                        }


                        onChange={(e) =>

                          handleRowChange(

                            index,

                            "pb_no",

                            e.target.value

                          )

                        }

                      />

                    </td>


                    <td>

                      <input

                        type="number"

                        placeholder="0.00"

                        value={
                          row.expenditure
                        }


                        onChange={(e) =>

                          handleRowChange(

                            index,

                            "expenditure",

                            e.target.value

                          )

                        }

                      />

                    </td>


                    <td>

                      <input

                        type="text"

                        placeholder="Narration"

                        value={
                          row.narration
                        }


                        onChange={(e) =>

                          handleRowChange(

                            index,

                            "narration",

                            e.target.value

                          )

                        }

                      />

                    </td>


                  </tr>

                )

              )}


            </tbody>


          </table>


        </div>


      </div>


      {/* =====================================================
          BOTTOM SECTION
      ===================================================== */}

      <div className="bottom-section">


        <div className="details-card">


          <h3>
            Transaction Details
          </h3>


          <div className="details-grid">


            <div className="field">


              <label>
                Bill No
              </label>


              <input

                type="text"

                value={
                  billNo
                }

                readOnly

              />


            </div>


            <div className="field">


              <label>
                Amount
              </label>


              <input

                type="number"

                placeholder="Amount"

                value={
                  amount
                }


                onChange={(e) =>

                  setAmount(
                    e.target.value
                  )

                }

              />


            </div>


          </div>


        </div>


        <div className="total-card">


          <h3>
            Running Total
          </h3>


          <input

            type="text"

            value={
              runningTotal
            }

            readOnly

            className="running-total"

          />


        </div>


      </div>


      {/* =====================================================
          CONTROLS
      ===================================================== */}

      <div className="controls-section">


        <button

          className="save-small-btn"

          onClick={
            handleSave
          }

        >

          Save

        </button>


        <button

          className="cheque-btn"

        >

          Cheque / DD Info

        </button>


      </div>


      {/* =====================================================
          ACTION BUTTONS
      ===================================================== */}

      <div className="action-buttons">


        <button

          className="calculate-btn"

          onClick={
            handleCalculate
          }

        >

          Calculate

        </button>


        <button

          className="save-btn"

          onClick={
            handleSave
          }

        >

          Save

        </button>


        <button

          className="clear-btn"

          onClick={
            handleClear
          }

        >

          Clear

        </button>


        <button

          className="print-btn"

          onClick={() =>
            window.print()
          }

        >

          Print

        </button>


      </div>


    </div>

  );

}


export default FourColumnDayBook;
